File: python/src/actr_harness/services/neuro_core/action_resolver.py
import logging
from dataclasses import asdict, dataclass, field
from typing import Any

from actr_harness.generated.grpc.actr import (
    BufferOperation,
    BufferState,
    ModuleSchema,
    NeuroAction,
)
from actr_harness.generated.grpc.actr.services import DecodeActionResponse
from actr_harness.observability import log_event, observe_boundary
from actr_harness.utils import dict_to_struct

from .buffers_view import BuffersView
from ..llm_client import LLMClient

logger = logging.getLogger(__name__)


@dataclass
class SemanticEntry:
    target_module_id: str
    command: str
    existing_params: dict[str, Any]
    semantic: dict[str, Any]
    meta: dict[str, Any]
    command_schema: str
    semantic_sources: dict[str, Any] = field(default_factory=dict)
    semantic_param_leaves: dict[str, str] = field(default_factory=dict)


class ActionResolver:
    def __init__(self, llm_client: LLMClient):
        self._llm_client = llm_client

    @observe_boundary("action_resolver.decode_action")
    async def decode_action(
            self,
            action_intent: NeuroAction,
            current_states: list[BufferState],
            schemas: list[ModuleSchema],
    ) -> DecodeActionResponse:
        view = BuffersView(current_states)
        keyed_schemas: dict[str, dict[str, str]] = {s.module_id: s.command_schemas for s in schemas}

        command_semantics: dict[str, dict[str, Any]] = {}
        neuro_intents: list[dict[str, Any]] = []
        meta_instructions: dict[str, dict[str, Any]] = {}

        for sem_key, sem_struct in action_intent.semantics.items():
            sem_dict = sem_struct.to_dict()
            prefix, _, sem_name = sem_key.partition(":")
            if prefix == "meta":
                meta_instructions[sem_name] = sem_dict
            elif prefix == "neuro":
                neuro_intents.append(sem_dict)
            elif prefix == "command":
                command_semantics[sem_name] = sem_dict

        determined_ops: list[BufferOperation] = []
        semantic_entries: list[SemanticEntry] = []

        for alias, base_op in action_intent.commands.items():
            base_params = self._resolve_placeholder_params(base_op.params.to_dict(), view)
            semantic = command_semantics.get(alias)

            if semantic is not None or len(meta_instructions) > 0:
                schema = keyed_schemas.get(base_op.target_module_id, {}).get(base_op.command)
                if schema is None:
                    raise ValueError(
                        "Missing schema for target_module_id="
                        f"'{base_op.target_module_id}', command='{base_op.command}', alias='{alias}'."
                    )

                sources_list: dict[str, Any] = {}
                param_leaves: dict[str, str] = {}
                if semantic is not None:
                    sources_raw = semantic.get("sources", [])
                    sources_list = {src: view.get(src) for src in sources_raw}
                    param_leaves = self._flatten_semantic_params(semantic.get("params", {}))

                semantic_entries.append(
                    SemanticEntry(
                        target_module_id=base_op.target_module_id,
                        command=base_op.command,
                        existing_params=base_params,
                        semantic=semantic if semantic is not None else {},
                        meta=meta_instructions,
                        command_schema=schema,
                        semantic_sources=sources_list,
                        semantic_param_leaves=param_leaves,
                    )
                )
            else:
                determined_ops.append(
                    BufferOperation(
                        target_module_id=base_op.target_module_id,
                        command=base_op.command,
                        params=dict_to_struct(base_params),
                    )
                )

        log_event(
            logger,
            logging.DEBUG,
            "action_decode.plan_built",
            "Built action decode plan.",
            rule_id=action_intent.rule_id,
            command_aliases=list(action_intent.commands.keys()),
            semantic_keys=list(action_intent.semantics.keys()),
            determined_operations=[_operation_summary(op) for op in determined_ops],
            semantic_entries=[_semantic_entry_summary(entry) for entry in semantic_entries],
            neuro_intent_count=len(neuro_intents),
        )

        if len(semantic_entries) > 0:
            semantic_ops = await self._llm_resolve_semantic_commands(
                determined_ops,
                semantic_entries,
                keyed_schemas,
            )
            determined_ops.extend(semantic_ops)
            log_event(
                logger,
                logging.INFO,
                "action_decode.semantic_resolution",
                "Resolved semantic command supplements.",
                semantic_entry_count=len(semantic_entries),
                produced_operations=[_operation_summary(op) for op in semantic_ops],
            )

        if len(neuro_intents) > 0:
            command_supplements = [
                {
                    "target_module_id": op.target_module_id,
                    "command": op.command,
                    "existing_params": op.params.to_dict(),
                }
                for op in determined_ops
            ]
            llm_ops = await self._llm_decode_fuzzy(
                command_supplements,
                neuro_intents,
                [
                    {"module_id": bs.module_id, "data": bs.data.to_dict()}
                    for bs in current_states
                ],
                keyed_schemas,
            )
            determined_ops.extend(llm_ops)
            log_event(
                logger,
                logging.INFO,
                "action_decode.neuro_resolution",
                "Resolved neuro intents into buffer operations.",
                neuro_intent_count=len(neuro_intents),
                produced_operations=[_operation_summary(op) for op in llm_ops],
            )

        log_event(
            logger,
            logging.INFO,
            "action_decode.completed",
            "Completed action decode.",
            rule_id=action_intent.rule_id,
            operation_count=len(determined_ops),
            operations=[_operation_summary(op) for op in determined_ops],
        )
        return DecodeActionResponse(operations=determined_ops)

    @observe_boundary("action_resolver.llm_resolve_semantic_commands")
    async def _llm_resolve_semantic_commands(
            self,
            determined_ops: list[BufferOperation],
            semantic_entries: list[SemanticEntry],
            keyed_schemas: dict[str, dict[str, str]],
    ) -> list[BufferOperation]:
        prompt_data = {
            "determined_ops": [_operation_summary(op) for op in determined_ops],
            "semantic_commands": [asdict(entry) for entry in semantic_entries],
            "module_schemas": keyed_schemas,
        }

        system_prompt = (
            "You are given already determined operations (do NOT include them in your output) "
            "and a set of incomplete semantic commands with parameters described in natural language. "
            "For each semantic command, resolve it into zero or more concrete operations according to "
            "its semantic description and any meta policies (e.g., skip if required sources are missing). "
            "Return ONLY the operations derived from the semantic commands (the already determined ones "
            "will be kept automatically). "
            "Output a strict JSON array of objects with keys: target_module_id, command, params. "
            "No extra text."
        )

        response = await self._llm_client.chat_json(prompt_data, system_prompt)
        ops: list[BufferOperation] = []
        if not isinstance(response, list):
            log_event(
                logger,
                logging.WARNING,
                "action_decode.semantic_resolution_invalid_payload",
                "Semantic command resolution returned a non-list payload.",
                payload_type=type(response).__name__,
            )
            return ops

        for item in response:
            if not isinstance(item, dict):
                log_event(
                    logger,
                    logging.DEBUG,
                    "action_decode.semantic_resolution_item_skipped",
                    "Skipped semantic command item with invalid type.",
                    payload_type=type(item).__name__,
                )
                continue

            try:
                ops.append(
                    BufferOperation(
                        target_module_id=item["target_module_id"],
                        command=item["command"],
                        params=dict_to_struct(item.get("params", {})),
                    )
                )
            except (KeyError, TypeError):
                log_event(
                    logger,
                    logging.DEBUG,
                    "action_decode.semantic_resolution_item_skipped",
                    "Skipped semantic command item with invalid shape.",
                    item=item,
                )
        return ops

    def _resolve_placeholder_params(self, value: Any, view: BuffersView) -> Any:
        if isinstance(value, str):
            if value.startswith("${") and value.endswith("}"):
                resolved = view.get(value[2:-1])
                return resolved if resolved is not None else value
            return value
        if isinstance(value, dict):
            return {key: self._resolve_placeholder_params(inner, view) for key, inner in value.items()}
        if isinstance(value, list):
            return [self._resolve_placeholder_params(inner, view) for inner in value]
        return value

    def _flatten_semantic_params(self, value: Any, prefix: str = "") -> dict[str, str]:
        if isinstance(value, dict):
            leaves: dict[str, str] = {}
            for key, child in value.items():
                path = f"{prefix}.{key}" if len(prefix) > 0 else key
                for leaf_key, leaf_value in self._flatten_semantic_params(child, path).items():
                    leaves[leaf_key] = leaf_value
            return leaves
        if isinstance(value, str):
            return {prefix: value}
        raise ValueError(
            f"Semantic param leaf at '{prefix or '<root>'}' must be a string, "
            f"got {type(value).__name__}."
        )

    @observe_boundary("action_resolver.llm_decode_fuzzy")
    async def _llm_decode_fuzzy(
            self,
            command_supplements: list[dict[str, Any]],
            neuro_intents: list[dict[str, Any]],
            buffers: list[dict[str, Any]],
            schemas: dict[str, Any],
    ) -> list[BufferOperation]:
        prompt_data = {
            "buffers": buffers,
            "module_schemas": schemas,
            "partial_commands": command_supplements,
            "neural_intents": neuro_intents,
        }
        system_prompt = (
            "Translate partial commands and neural intents into concrete operations. "
            "Each operation must use a valid module_id from schemas, a command defined there, "
            "and parameters with correct types. "
            "Output a strict JSON array of objects with keys: target_module_id, command, params. "
            "No commentary."
        )
        ops_raw = await self._llm_client.chat_json(prompt_data, system_prompt)
        ops: list[BufferOperation] = []
        if not isinstance(ops_raw, list):
            log_event(
                logger,
                logging.WARNING,
                "action_decode.neuro_resolution_invalid_payload",
                "Fuzzy decode returned a non-list payload.",
                payload_type=type(ops_raw).__name__,
            )
            return ops

        for item in ops_raw:
            try:
                ops.append(
                    BufferOperation(
                        target_module_id=item["target_module_id"],
                        command=item["command"],
                        params=dict_to_struct(item.get("params", {})),
                    )
                )
            except (KeyError, TypeError):
                log_event(
                    logger,
                    logging.DEBUG,
                    "action_decode.neuro_resolution_item_skipped",
                    "Skipped fuzzy decode item with invalid shape.",
                    item=item,
                )
        return ops


def _operation_summary(operation: BufferOperation) -> dict[str, Any]:
    return {
        "target_module_id": operation.target_module_id,
        "command": operation.command,
        "params": operation.params.to_dict(),
    }


def _semantic_entry_summary(entry: SemanticEntry) -> dict[str, Any]:
    return {
        "target_module_id": entry.target_module_id,
        "command": entry.command,
        "semantic_sources": entry.semantic_sources,
        "semantic_param_leaves": entry.semantic_param_leaves,
        "meta": entry.meta,
    }

File: Dotnet/src/Harness.Core/NeuroCore/ActionResolver.cs
using System.Text.Json.Nodes;
using Harness.Abstractions.Actr;
using Harness.Shared.Observability;
using Microsoft.Extensions.Logging;

namespace Harness.Core.NeuroCore;

public class ActionResolver(LlmClient llmClient, ILogger<ActionResolver> logger) : IProvideLogger
{
    private sealed record SemanticEntry(
        string TargetModuleId,
        string Command,
        Dictionary<string, object?> ExistingParams,
        Dictionary<string, object?> Semantic,
        Dictionary<string, Dictionary<string, object?>> Meta,
        string CommandSchema,
        Dictionary<string, object?> SemanticSources,
        Dictionary<string, string> SemanticParamLeaves);

    public ILogger Logger => logger;

    [TraceSpan]
    public async Task<IReadOnlyList<BufferOperation>> DecodeActionAsync(
        NeuroAction actionIntent,
        IReadOnlyList<BufferState> currentStates,
        IReadOnlyList<ModuleSchema> schemas,
        CancellationToken cancellationToken = default)
    {
        var view = new BuffersView(currentStates);
        var keyedSchemas = schemas.ToDictionary(
            schema => schema.ModuleId,
            schema => schema.CommandSchemas.ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.Ordinal),
            StringComparer.Ordinal);

        var commandSemantics = new Dictionary<string, Dictionary<string, object?>>(StringComparer.Ordinal);
        var neuroIntents = new List<Dictionary<string, object?>>();
        var metaInstructions = new Dictionary<string, Dictionary<string, object?>>(StringComparer.Ordinal);

        foreach (var semanticsEntry in actionIntent.Semantics)
        {
            var parts = semanticsEntry.Key.Split(':', 2);
            if (parts.Length != 2)
            {
                continue;
            }

            var semantics = ProtobufDataConverter.ToPlainObjectMap(semanticsEntry.Value);
            switch (parts[0])
            {
                case "meta":
                    metaInstructions[parts[1]] = semantics;
                    break;
                case "neuro":
                    neuroIntents.Add(semantics);
                    break;
                case "command":
                    commandSemantics[parts[1]] = semantics;
                    break;
            }
        }

        var determinedOperations = new List<BufferOperation>();
        var semanticEntries = new List<SemanticEntry>();

        foreach (var (alias, baseOperation) in actionIntent.Commands)
        {
            var baseParams = ResolvePlaceholderParams(
                ProtobufDataConverter.ToPlainObjectMap(baseOperation.Params),
                view);
            var semantic = commandSemantics.GetValueOrDefault(alias);

            if (semantic is not null || metaInstructions.Count > 0)
            {
                if (!keyedSchemas.TryGetValue(baseOperation.TargetModuleId, out var moduleSchemas)
                    || !moduleSchemas.TryGetValue(baseOperation.Command, out var schema))
                {
                    throw new InvalidOperationException(
                        $"Missing schema for target_module_id='{baseOperation.TargetModuleId}', command='{baseOperation.Command}', alias='{alias}'.");
                }

                var sources = new Dictionary<string, object?>(StringComparer.Ordinal);
                var semanticParamLeaves = new Dictionary<string, string>(StringComparer.Ordinal);
                if (semantic is not null)
                {
                    foreach (var source in ReadSources(semantic))
                    {
                        sources[source] = view.Get(source);
                    }

                    if (semantic.TryGetValue("params", out var paramsValue))
                    {
                        semanticParamLeaves = FlattenSemanticParams(paramsValue);
                    }
                }

                semanticEntries.Add(new SemanticEntry(
                    baseOperation.TargetModuleId,
                    baseOperation.Command,
                    baseParams,
                    semantic ?? new Dictionary<string, object?>(StringComparer.Ordinal),
                    metaInstructions,
                    schema,
                    sources,
                    semanticParamLeaves));
            }
            else
            {
                determinedOperations.Add(new BufferOperation
                {
                    TargetModuleId = baseOperation.TargetModuleId,
                    Command = baseOperation.Command,
                    Params = ProtobufStructConverter.ToStruct(baseParams)
                });
            }
        }

        if (semanticEntries.Count > 0)
        {
            var semanticOperations = await ResolveSemanticCommandsAsync(
                determinedOperations,
                semanticEntries,
                keyedSchemas,
                cancellationToken);
            determinedOperations.AddRange(semanticOperations);
        }

        if (neuroIntents.Count <= 0) return determinedOperations;

        var neuroOperations = await DecodeNeuroIntentsAsync(
            determinedOperations,
            neuroIntents,
            currentStates,
            keyedSchemas,
            cancellationToken);
        determinedOperations.AddRange(neuroOperations);

        return determinedOperations;
    }

    [TraceSpan]
    private async Task<IReadOnlyList<BufferOperation>> ResolveSemanticCommandsAsync(
        IReadOnlyList<BufferOperation> determinedOperations,
        IReadOnlyList<SemanticEntry> semanticEntries,
        IReadOnlyDictionary<string, Dictionary<string, string>> keyedSchemas,
        CancellationToken cancellationToken)
    {
        var promptData = new Dictionary<string, object?>
        {
            ["determined_ops"] = determinedOperations.Select(OperationSummary).ToList(),
            ["semantic_commands"] = semanticEntries.Select(entry => new Dictionary<string, object?>
            {
                ["target_module_id"] = entry.TargetModuleId,
                ["command"] = entry.Command,
                ["existing_params"] = entry.ExistingParams,
                ["semantic"] = entry.Semantic,
                ["meta"] = entry.Meta,
                ["command_schema"] = entry.CommandSchema,
                ["semantic_sources"] = entry.SemanticSources,
                ["semantic_param_leaves"] = entry.SemanticParamLeaves
            }).ToList(),
            ["module_schemas"] = keyedSchemas
        };

        const string systemPrompt =
            "You are given already determined operations (do NOT include them in your output) "
            + "and a set of incomplete semantic commands with parameters described in natural language. "
            + "For each semantic command, resolve it into zero or more concrete operations according to "
            + "its semantic description and any meta policies (e.g., skip if required sources are missing). "
            + "Return ONLY the operations derived from the semantic commands (the already determined ones "
            + "will be kept automatically). "
            + "Output a strict JSON array of objects with keys: target_module_id, command, params. "
            + "No extra text.";

        var response = await llmClient.ChatJsonAsync(promptData, systemPrompt, cancellationToken);
        return ParseOperations(response);
    }

    [TraceSpan]
    private async Task<IReadOnlyList<BufferOperation>> DecodeNeuroIntentsAsync(
        IReadOnlyList<BufferOperation> determinedOperations,
        IReadOnlyList<Dictionary<string, object?>> neuroIntents,
        IReadOnlyList<BufferState> currentStates,
        IReadOnlyDictionary<string, Dictionary<string, string>> keyedSchemas,
        CancellationToken cancellationToken)
    {
        var promptData = new Dictionary<string, object?>
        {
            ["buffers"] = currentStates.Select(state => new Dictionary<string, object?>
            {
                ["module_id"] = state.ModuleId,
                ["data"] = ProtobufDataConverter.ToPlainObjectMap(state.Data)
            }).ToList(),
            ["module_schemas"] = keyedSchemas,
            ["partial_commands"] = determinedOperations.Select(operation => new Dictionary<string, object?>
            {
                ["target_module_id"] = operation.TargetModuleId,
                ["command"] = operation.Command,
                ["existing_params"] = ProtobufDataConverter.ToPlainObjectMap(operation.Params)
            }).ToList(),
            ["neural_intents"] = neuroIntents
        };

        const string systemPrompt =
            "Translate partial commands and neural intents into concrete operations. "
            + "Each operation must use a valid module_id from schemas, a command defined there, "
            + "and parameters with correct types. "
            + "Output a strict JSON array of objects with keys: target_module_id, command, params. "
            + "No commentary.";

        var response = await llmClient.ChatJsonAsync(promptData, systemPrompt, cancellationToken);
        return ParseOperations(response);
    }

    private static Dictionary<string, object?> ResolvePlaceholderParams(
        IReadOnlyDictionary<string, object?> value,
        BuffersView view)
    {
        return value.ToDictionary(pair => pair.Key, pair => ResolvePlaceholderValue(pair.Value, view),
            StringComparer.Ordinal);
    }

    private static object? ResolvePlaceholderValue(object? value, BuffersView view)
    {
        return value switch
        {
            string text when text.StartsWith("${", StringComparison.Ordinal) &&
                             text.EndsWith("}", StringComparison.Ordinal) => view.Get(text[2..^1]) ?? text,
            IReadOnlyDictionary<string, object?> map => map.ToDictionary(pair => pair.Key,
                pair => ResolvePlaceholderValue(pair.Value, view), StringComparer.Ordinal),
            IDictionary<string, object?> map => map.ToDictionary(pair => pair.Key,
                pair => ResolvePlaceholderValue(pair.Value, view), StringComparer.Ordinal),
            IEnumerable<object?> list => list.Select(item => ResolvePlaceholderValue(item, view)).ToList(),
            _ => value
        };
    }

    private static Dictionary<string, string> FlattenSemanticParams(object? value, string prefix = "")
    {
        while (true)
        {
            switch (value)
            {
                case IReadOnlyDictionary<string, object?> map:
                {
                    var leaves = new Dictionary<string, string>(StringComparer.Ordinal);
                    foreach (var entry in map)
                    {
                        var path = string.IsNullOrEmpty(prefix) ? entry.Key : $"{prefix}.{entry.Key}";
                        foreach (var leaf in FlattenSemanticParams(entry.Value, path))
                        {
                            leaves[leaf.Key] = leaf.Value;
                        }
                    }

                    return leaves;
                }
                case IDictionary<string, object?> map:
                    value = new Dictionary<string, object?>(map, StringComparer.Ordinal);
                    continue;
                case string leaf:
                    return new Dictionary<string, string>(StringComparer.Ordinal) { [prefix] = leaf };
                default:
                    throw new InvalidOperationException(
                        $"Semantic param leaf at '{(string.IsNullOrEmpty(prefix) ? "<root>" : prefix)}' must be a string, got {value?.GetType().Name ?? "null"}.");
            }
        }
    }

    private static IEnumerable<string> ReadSources(Dictionary<string, object?> semantic)
    {
        if (!semantic.TryGetValue("sources", out var sourcesValue))
        {
            return [];
        }

        return sourcesValue switch
        {
            IEnumerable<object?> list => list.OfType<string>(),
            string single => [single],
            _ => []
        };
    }

    private static IReadOnlyList<BufferOperation> ParseOperations(JsonNode? response)
    {
        if (response is not JsonArray array)
        {
            return [];
        }

        var operations = new List<BufferOperation>();
        foreach (var item in array)
        {
            if (item is null ||
                ProtobufStructConverter.JsonNodeToPlainObject(item) is not Dictionary<string, object?> operationMap)
            {
                continue;
            }

            if (!operationMap.TryGetValue("target_module_id", out var targetModuleIdValue)
                || targetModuleIdValue is not string targetModuleId
                || !operationMap.TryGetValue("command", out var commandValue)
                || commandValue is not string command)
            {
                continue;
            }

            var plainParams = operationMap.TryGetValue("params", out var paramsValue)
                              && paramsValue is Dictionary<string, object?> paramsMap
                ? paramsMap
                : new Dictionary<string, object?>(StringComparer.Ordinal);

            operations.Add(new BufferOperation
            {
                TargetModuleId = targetModuleId,
                Command = command,
                Params = ProtobufStructConverter.ToStruct(plainParams)
            });
        }

        return operations;
    }

    private static Dictionary<string, object?> OperationSummary(BufferOperation operation)
    {
        return new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["target_module_id"] = operation.TargetModuleId,
            ["command"] = operation.Command,
            ["params"] = ProtobufDataConverter.ToPlainObjectMap(operation.Params)
        };
    }
}